cloud/scope/clients.go

```diff
@@ -37,40 +38,68 @@ const (
 
 // AzureClients contains all the Azure clients used by the scopes.
 type AzureClients struct {
-	SubscriptionID             string
+	Authorizer                 autorest.Authorizer
+	environment                string
 	ResourceManagerEndpoint    string
 	ResourceManagerVMDNSSuffix string
-	Authorizer                 autorest.Authorizer
+	SubscriptionID             string
+	tenantID                   string
+	clientID                   string
+	clientSecret               string
+}
+
+// CloudEnvironment returns the Azure environment the controller runs in.
+func (c *AzureClients) CloudEnvironment() string {
+	return c.environment
+}
+
+// TenantID returns the Azure tenants the controller runs in.
+func (c *AzureClients) TenantID() string {
+	return c.tenantID
+}
+
+// ClientID returns the Azure environment the controller runs in.
+func (c *AzureClients) ClientID() string {
+	return c.clientID
+}
+
+// ClientSecret returns the Azure tenants the controller runs in.
+func (c *AzureClients) ClientSecret() string {
+	return c.clientSecret
 }
 
 func (c *AzureClients) setCredentials(subscriptionID string) error {
-	subID, err := getSubscriptionID(subscriptionID)
-	if err != nil {
-		return err
-	}
-	c.SubscriptionID = subID
 	settings, err := auth.GetSettingsFromEnvironment()
 	if err != nil {
 		return err
 	}
+
+	if subscriptionID == "" {
+		subscriptionID = settings.GetSubscriptionID()
+		if subscriptionID == "" {
+			return fmt.Errorf("error creating azure services. subscriptionID is not set in cluster or AZURE_SUBSCRIPTION_ID env var")
+		}
+	}
+
+	c.SubscriptionID = subscriptionID
+	c.tenantID = strings.TrimSuffix(settings.Values[auth.TenantID], "\n")
+	c.clientID = strings.TrimSuffix(settings.Values[auth.ClientID], "\n")
+	c.clientSecret = strings.TrimSuffix(settings.Values[auth.ClientSecret], "\n")
+
+	c.environment = settings.Values[auth.EnvironmentName]
+	if c.environment == "" {
+		c.environment = azure.PublicCloud.Name
+	}
+
 	c.ResourceManagerEndpoint = settings.Environment.ResourceManagerEndpoint
 	c.ResourceManagerVMDNSSuffix = GetAzureDNSZoneForEnvironment(settings.Environment.Name)
 	settings.Values[auth.SubscriptionID] = subscriptionID
+	settings.Values[auth.TenantID] = c.tenantID
```

Comment:
why is this set here? Doesn't the value of c.tenantID come from `strings.TrimSuffix(settings.Values[auth.TenantID], "\n")` a few lines above?

Reply:
you need to re-apply it into settings so `settings.GetAuthorizer()` works properly -___-

we should split all the env settings and credentials code up. With cluster describer we'll mostly split out subscription/credentials, should make sure we have a nice place for tenant ID (even if it's just the env for now)